extensions/geometry/init.lua

```diff
@@ -1,44 +1,494 @@
+---    * `"X1 Y1 > X2 Y2"` or `"X1,Y1,X2,Y2"` (or variations thereof) creates a rect given two opposite corners
+---
+--- You can use any of these anywhere an hs.geometry object is expected in Hammerspoon; the constructor will be called for you.
 
-local geometry = require "hs.geometry.internal"
+-- remove copy-on-new
+--  * another `hs.geometry` object - creates a copy of that object
 
---- hs.geometry.rotateCCW(point, aroundpoint, ntimes = 1) -> point
---- Function
---- Rotates a point around another point N times
+--TODO allow segments? (like rects, but no norm())
+
+local rawget,rawset,type,pairs,tonumber,sqrt=rawget,rawset,type,pairs,tonumber,math.sqrt
+local min,max,atan,smatch,sformat=math.min,math.max,math.atan,string.match,string.format
+local getmetatable,setmetatable=getmetatable,setmetatable
+
+local geometry = {}--require "hs.geometry.internal"
+
+local ws=' *'
+local sepout=ws..'([ ,>/])'..ws
+local sepin=ws..'([ x*,])'..ws
+local number='(%-?%d*%.?%d*)'
+local function parse(s)
+  local x,y,w,h
+  local sept,n3,sep2,n4
+  local n1,sep1,n2,i = smatch(s,number..sepin..number..'()')
+  if i then sept,n3,sep2,n4 = smatch(s,sepout..number..sepin..number,i) end
+  if sep1=='x' or sep1=='*' then --it's a size
+    w=tonumber(n1) h=tonumber(n2)
+    if not w then error('Cannot parse width',3) end
+    if not h then error('Cannot parse heigth',3) end
+  else
+    x=tonumber(n1) y=tonumber(n2)
+    if not x then error('Cannot parse x coord',3) end
+    if not y then error('Cannot parse y coord',3) end
+    if sept=='>' or sep2==' ' or sep2==',' then --it's a rectx2y2
+      w=tonumber(n3)-x h=tonumber(n4)-y
+      if not w then error('Cannot parse x2 coord',3) end
+      if not h then error('Cannot parse y2 coord',3) end
+    elseif sep2 then --it's a rectsize
+      w=tonumber(n3) h=tonumber(n4)
+      if not w then error('Cannot parse width',3) end
+      if not h then error('Cannot parse heigth',3) end
+    end
+  end
+  return x,y,w,h
+end
+
+local function gettype(t)
+  if t._x and t._y then
+    if t._w and t._h then return 'rect' else return 'point' end
+  elseif t._w and t._h then return 'size' end
+end
+local function norm(t)
+  if t._x and t._w and t._w<0 then t._x=t._x+t._w t._w=-t._w end
+  if t._y and t._h and t._h<0 then t._y=t._y+t._h t._h=-t._h end
+end
+
+local function new(x,y,w,h)
+  if getmetatable(x)==geometry then return x end -- disable copy-on-new
+  if type(x)=='string' then
+    x,y,w,h=parse(x)
+  elseif type(x)=='table' then
+    if x[1] then x,y,w,h=x[1],x[2],x[3],x[4]
+    else
+      local t=x
+      x,y=t.x or t.x1,t.y or t.y1
+      w,h=t.w or t.x2-x,t.h or t.y2-y
+    end
+  end
+  local t={_x=x,_y=y,_w=w,_h=h}
+  for k,v in pairs(t) do if type(v)~='number' then t[k]=nil end end
+  norm(t)
+  if not gettype(t) then error('Cannot create geometry object, wrong arguments',2) end
+  return setmetatable(t,geometry)
+end
+--- hs.geometry.new(...) -> hs.geometry
```

Comment:
I'm not a fan of having "hs.geometry" as the return type. We're not very consistent about this yet, but I think the word "object" needs to be in there - maybe "geometry object"?